In [1]:
%cd TestFinanceFormula

[Errno 2] No such file or directory: 'TestFinanceFormula'
/Users/trancatkhanh/Desktop/TestFinanceFormula


In [2]:
from generator import Generator
import pandas as pd
import numpy as np
data = pd.read_excel("HOSE_File3_2023_Field.xlsx")
data = data[data["TIME"] >= 2007]
data = data[data["TIME"] <= 2022]
data

SYMBOL EXCHANGE  TIME    PROFIT      VALUEARG  Cash and cash equivalents  \
339     AAA     HOSE  2022  0.476440  1.559239e+11                   0.318878   
340     AAM     HOSE  2022  0.744966  1.660500e+09                   0.396938   
341     AAT     HOSE  2022  0.410448  8.211400e+09                   0.109150   
342     ABR     HOSE  2022  0.475490  2.440000e+07                   0.120767   
343     ABS     HOSE  2022  0.255578  2.783360e+10                   0.001594   
...     ...      ...   ...       ...           ...                        ...   
4721    VPK     HOSE  2007  0.502778  1.682800e+09                   0.020743   
4722    VSH     HOSE  2007  0.455263  1.280530e+10                   0.005633   
4723    VTA     HOSE  2007  0.548485  2.194500e+09                   0.065528   
4724    VTB     HOSE  2007  0.506849  8.179000e+08                   0.032938   
4725    VTC     HOSE  2007  0.531832  5.777000e+08                   0.070280   

      Short Term Financial Investments  Short term receivables  \
339                           0.069954                0.287971   
340                           0.039397                0.049874   
341                           0.001170                0.247543   
342                           0.354600                0.102462   
343                           0.172515                0.465351   
...                                ...                     ...   
4721                          0.000000                0.057294   
4722                          0.006091                0.002983   
4723                          0.000000                0.497201   
4724                          0.024943                0.120339   
4725                          0.001582                0.295638   

      Total Inventories  Total Other Current Assets  ...  \
339            0.159968                    0.022042  ...   
340            0.544791                    0.016006  ...   
341            0.020843                    0.022754  ...   
342            0.001000                    0.000924  ...   
343            0.010245                    0.000452  ...   
...                 ...                         ...  ...   
4721           0.116012                    0.024933  ...   
4722           0.002712                    0.000119  ...   
4723           0.481074                    0.000587  ...   
4724           0.059402                    0.018610  ...   
4725           0.116344                    0.002460  ...   

      General and administration expenses  Net operating profit  Other income  \
339                              0.031236              0.064973      0.001382   
340                              0.041257             -0.006316      0.032221   
341                              0.015443              0.038660      0.007479   
342                              0.054159              0.056796      0.001087   
343                              0.004550              0.041596      0.000842   
...                                   ...                   ...           ...   
4721                             0.026945              0.031980      0.002445   
4722                             0.000696              0.028461      0.000139   
4723                             0.021797              0.016901      0.002553   
4724                             0.016983              0.038285      0.009579   
4725                             0.043458              0.038562      0.002154   

      Other expenses  Other profit  Total accounting profit before tax  \
339         0.002542     -0.001160                            0.063813   
340         0.024444      0.007777                            0.001461   
341         0.003214      0.004265                            0.042925   
342         0.000356      0.000731                            0.057526   
343         0.000861     -0.000019                            0.041576   
...              ...           ...                                 ...   
4721        0.000467      0

# Thực hiện Test code công thức : Ngoài hệ thống của anh An

In [17]:
def get_data(data,years_begin,years_end):
    symbols_end = data[data["TIME"] == int(years_begin)]["SYMBOL"].unique()
    symbols = []
    for symbol in symbols_end:
        exists_in_all_years = True
        for year in list(range(int(years_begin), int(years_end+1))):
            if symbol not in data[data["TIME"] == year]["SYMBOL"].unique():
                exists_in_all_years = False
                break
        if exists_in_all_years:
            symbols.append(symbol)
    data = data[data["SYMBOL"].isin(symbols)]
    # print the data from the years_begin to years_end in the TIME columns
    data = data[data["TIME"] >= int(years_begin)]
    data = data[data["TIME"] <= int(years_end)]
    return data
filtered_data=  get_data(data,2007,2022)


**Các công ty phải tồn tại xuyên suốt giai đoạn năm 2007 - 2022**

In [112]:
filtered_data.head()

SYMBOL EXCHANGE  TIME    PROFIT      VALUEARG  Cash and cash equivalents  \
367    BMC     HOSE  2022  0.585062  4.483900e+09                   0.266337   
368    BMP     HOSE  2022  1.006902  4.236900e+09                   0.032402   
382    CII     HOSE  2022  0.412308  3.520647e+11                   0.087449   
384    CLC     HOSE  2022  0.959108  1.197900e+09                   0.058603   
391    COM     HOSE  2022  0.480028  4.280000e+08                   0.083097   

     Short Term Financial Investments  Short term receivables  \
367                          0.000000                0.015456   
368                          0.152698                0.069265   
382                          0.083757                0.555696   
384                          0.000000                0.191794   
391                          0.015264                0.018864   

     Total Inventories  Total Other Current Assets  ...  \
367           0.303222                    0.046970  ...   
368           0.115955                    0.005905  ...   
382           0.577221                    0.027793  ...   
384           0.694734                    0.007103  ...   
391           0.098535                    0.006660  ...   

     General and administration expenses  Net operating profit  Other income  \
367                             0.045098              0.072318      0.000865   
368                             0.016407              0.050926      0.000467   
382                             0.062709             -0.009451      0.003761   
384                             0.071940              0.159749      0.000729   
391                             0.013433              0.045497      0.003203   

     Other expenses  Other profit  Total accounting profit before tax  \
367    3.680582e-03     -0.002816                            0.069502   
368    1.143017e-03     -0.000676                            0.050250   
382    7.734962e-03     -0.003974                           -0.013425   
384    6.110503e-06      0.000723                            0.160473   
391    8.149771e-07      0.003202                            0.048698   

     Current income tax  Deferred income tax  Profit after tax  ValueARG  
367            0.011103             0.000000          0.058399  0.015013  
368            0.009020             0.001064          0.040166  0.000794  
382            0.026742            -0.009453         -0.030715  0.044670  
384            0.032412             0.000000          0.128060  0.001133  
391            0.009855            -0.000365          0.039208  0.000417  

[5 rows x 40 columns]

**Cho công thức A+B  vs A là cột Cash and cash equivalents vs B là cột Short Term Financial Investments**
**==> ta có bảng weight**

In [31]:
new_df = pd.DataFrame({'sum': filtered_data['Cash and cash equivalents'] + filtered_data['Short Term Financial Investments'], 'SYMBOL': filtered_data['SYMBOL'], 'TIME': filtered_data['TIME']})
new_df = new_df.reset_index(drop=True)
new_df

sum SYMBOL  TIME
0    0.266337    BMC  2022
1    0.185100    BMP  2022
2    0.171206    CII  2022
3    0.058603    CLC  2022
4    0.098361    COM  2022
..        ...    ...   ...
939  0.060203    VIP  2007
940  0.135642    VIS  2007
941  0.015428    VNM  2007
942  0.011723    VSH  2007
943  0.057881    VTB  2007

[944 rows x 3 columns]

**Weight của các năm từ 2007 đến 2021**

In [113]:
new_df_2007_2021 = new_df[new_df['TIME'] != 2022]
new_df_2007_2021 = new_df_2007_2021.reset_index(drop=True)
new_df_2007_2021.head()

sum SYMBOL  TIME
0  0.018117    BMC  2021
1  0.290072    BMP  2021
2  0.178566    CII  2021
3  0.057565    CLC  2021
4  0.109757    COM  2021

**Mean Weight của các năm từ 2007 đến 2021**

In [92]:
# create the new dataframe of all time except 2022
new_df_2007_2021 = new_df[new_df['TIME'] != 2022]
new_df_2007_2021 = new_df_2007_2021.reset_index(drop=True)
new_mean = new_df_2007_2021.groupby(['SYMBOL']).mean().drop(columns=['TIME'])
new_mean = new_mean.reset_index()
new_mean.columns = ['SYMBOL', 'mean']
new_mean.head()

SYMBOL      mean
0    BMC  0.141737
1    BMP  0.123226
2    CII  0.200663
3    CLC  0.110731
4    COM  0.177818

**Weight của năm 2022**

In [93]:
new_df_2022 = new_df[new_df['TIME'] == 2022]
new_df_2022 = new_df_2022.reset_index(drop=True)
new_df_2022.head()


sum SYMBOL  TIME
0  0.266337    BMC  2022
1  0.185100    BMP  2022
2  0.171206    CII  2022
3  0.058603    CLC  2022
4  0.098361    COM  2022

**Độ chêch lệch  theo phần trăm của năm 2022 so cái mean các năm trước đó**

In [96]:
# How to combine new_df_2022 and new_mean
new_df_2022_1 = pd.merge(new_df_2022, new_mean, on='SYMBOL')
new_df_2022_1.columns = ['sum','SYMBOL', 'TIME', 'mean']
new_df_2022_1['diff'] = ((new_df_2022_1['sum'] - new_df_2022_1['mean'])/new_df_2022_1['mean'])*100
new_df_2022_1['SUM'] = new_df_2022_1['sum'].astype(float)

new_df_2022_1 = new_df_2022_1.drop(columns=['sum'])
new_df_2022_1.head()


SYMBOL  TIME      mean       diff       SUM
0    BMC  2022  0.141737  87.909331  0.266337
1    BMP  2022  0.123226  50.211228  0.185100
2    CII  2022  0.200663 -14.679694  0.171206
3    CLC  2022  0.110731 -47.075970  0.058603
4    COM  2022  0.177818 -44.684162  0.098361

In [97]:

print(new_df_2022_1[new_df_2022_1['diff'] == new_df_2022_1['diff'].max()])
#Print the symbol has the max values of diff as string
print('Công ty đầu tư năm 2022 là ' ,new_df_2022_1[new_df_2022_1['diff'] == new_df_2022_1['diff'].max()]['SYMBOL'].values[0] )
print('Tỷ lệ chêch lệch là ' ,new_df_2022_1[new_df_2022_1['diff'] == new_df_2022_1['diff'].max()]['diff'].values[0] , '%')

   SYMBOL  TIME      mean       diff       SUM
53    VID  2022  0.187682  143.70706  0.457393
Công ty đầu tư năm 2022 là  VID
Tỷ lệ chêch lệch là  143.7070595958498 %


**Hàm tính ra công ty đầu tư**

In [109]:
def get_data(data,years_begin,years_end):
    symbols_end = data[data["TIME"] == int(years_begin)]["SYMBOL"].unique()
    symbols = []
    for symbol in symbols_end:
        exists_in_all_years = True
        for year in list(range(int(years_begin), int(years_end+1))):
            if symbol not in data[data["TIME"] == year]["SYMBOL"].unique():
                exists_in_all_years = False
                break
        if exists_in_all_years:
            symbols.append(symbol)
    data = data[data["SYMBOL"].isin(symbols)]
    # print the data from the years_begin to years_end in the TIME columns
    data = data[data["TIME"] >= int(years_begin)]
    data = data[data["TIME"] <= int(years_end)]
    return data


def find_company_to_invest(data,years_begin,years_end):
    filtered_data = get_data(data,years_begin,years_end)
    new_df = pd.DataFrame({'sum': filtered_data['Cash and cash equivalents'] + filtered_data['Short Term Financial Investments'], 'SYMBOL': filtered_data['SYMBOL'], 'TIME': filtered_data['TIME']})
    new_df = new_df.reset_index(drop=True)
    new_df_2007_2021 = new_df[new_df['TIME'] != years_end]
    new_df_2007_2021 = new_df_2007_2021.reset_index(drop=True)

    new_df_2022 = new_df[new_df['TIME'] == years_end]
    new_df_2022 = new_df_2022.reset_index(drop=True)

    new_mean = new_df_2007_2021.groupby(['SYMBOL']).mean().drop(columns=['TIME'])
    new_mean = new_mean.reset_index()
    new_mean.columns = ['SYMBOL', 'mean']
    new_df_2022_1 = pd.merge(new_df_2022, new_mean, on='SYMBOL')
    new_df_2022_1.columns = ['sum','SYMBOL', 'TIME', 'mean of previous years']
    new_df_2022_1['weight of year end'] = new_df_2022_1['sum'].astype(float)
    new_df_2022_1['diff in percentage'] = ((new_df_2022_1['sum'] - new_df_2022_1['mean of previous years'])/new_df_2022_1['mean of previous years'])*100
    new_df_2022_1 = new_df_2022_1.drop(columns=['sum'])
    print('Công ty đầu tư năm ' +str(years_end)+' là ' ,new_df_2022_1[new_df_2022_1['diff in percentage'] == new_df_2022_1['diff in percentage'].max()]['SYMBOL'].values[0] )
    print('Tỷ lệ chêch lệch là ' ,new_df_2022_1[new_df_2022_1['diff in percentage'] == new_df_2022_1['diff in percentage'].max()]['diff in percentage'].values[0] , '%')
    return new_df_2022_1
final_table =find_company_to_invest(data,2007,2009)

Công ty đầu tư năm 2009 là  TTP
Tỷ lệ chêch lệch là  3961.6086052321557 %


# Thực hiện Test code công thức : Merge Hệ thống của anh An

**Xử lý dữ liệu để**
**các công ty phải tồn tại xuyên suốt giai đoạn năm 2007 - 2022**

In [3]:
def get_data(data,years_begin,years_end):
    symbols_end = data[data["TIME"] == int(years_begin)]["SYMBOL"].unique()
    symbols = []
    for symbol in symbols_end:
        exists_in_all_years = True
        for year in list(range(int(years_begin), int(years_end+1))):
            if symbol not in data[data["TIME"] == year]["SYMBOL"].unique():
                exists_in_all_years = False
                break
        if exists_in_all_years:
            symbols.append(symbol)
    data = data[data["SYMBOL"].isin(symbols)]
    # print the data from the years_begin to years_end in the TIME columns
    data = data[data["TIME"] >= int(years_begin)]
    data = data[data["TIME"] <= int(years_end)]
    return data
filtered_data =  get_data(data,2007,2022)

In [4]:
vis = Generator(filtered_data,
                generation_method=0,
                 required_fields=[],
                 multiple_cycles=False,
                 investment_method=2,
                 interest_rate=1.06,
                 measurement_method=0,
                 target=1.4,
                 max_loop = 7,
                 numerator_condition = True
                 )

Các cột không được coi là biến: ['TIME', 'PROFIT', 'SYMBOL', 'VALUEARG', 'EXCHANGE']


In [ ]:
vis.generate("Data_CT", 100, 200000)

# Check Logic sau khi merge hệ thống



# Ví dụ  có công thức là +16 trong bảng excel formula_14_07_2023_16_15_00.csv là   +16,1.2169685235726446,2022,1.280645161290323

# Công thức +16 tức là lấy cột Sales

In [6]:
df = pd.read_csv('Data_CT/formula_14_07_2023_16_15_00.csv')
df.head(1)

formula  f_profit  TIME    profit
0       16  1.216969  2022  1.280645

In [7]:
# print sales columns
filtered_data[['Sales','TIME','SYMBOL']]


Sales  TIME SYMBOL
0    0.460201  2022    BMC
1    0.855284  2022    BMP
2    0.369054  2022    CII
3    2.025669  2022    CLC
4    2.595015  2022    COM
..        ...   ...    ...
939  0.266728  2007    VIP
940  2.215846  2007    VIS
941  0.221721  2007    VNM
942  0.045975  2007    VSH
943  0.773842  2007    VTB

[944 rows x 3 columns]

# Check lại bàng hàm find_company_to_invest với cột SALES nhận thấy năm 2022 nên đầu tư mã VSH 


In [8]:
def find_company_to_invest(data,years_begin,years_end):
    filtered_data = get_data(data,years_begin,years_end)
    new_df = pd.DataFrame({'sum': filtered_data['Sales'], 'SYMBOL': filtered_data['SYMBOL'], 'TIME': filtered_data['TIME']})
    new_df = new_df.reset_index(drop=True)
    new_df_2007_2021 = new_df[new_df['TIME'] != years_end]
    new_df_2007_2021 = new_df_2007_2021.reset_index(drop=True)

    new_df_2022 = new_df[new_df['TIME'] == years_end]
    new_df_2022 = new_df_2022.reset_index(drop=True)

    new_mean = new_df_2007_2021.groupby(['SYMBOL']).mean().drop(columns=['TIME'])
    new_mean = new_mean.reset_index()
    new_mean.columns = ['SYMBOL', 'mean']
    new_df_2022_1 = pd.merge(new_df_2022, new_mean, on='SYMBOL')
    new_df_2022_1.columns = ['sum','SYMBOL', 'TIME', 'mean of previous years']
    new_df_2022_1['weight of year end'] = new_df_2022_1['sum'].astype(float)
    new_df_2022_1['diff in percentage'] = ((new_df_2022_1['sum'] - new_df_2022_1['mean of previous years'])/new_df_2022_1['mean of previous years'])*100
    new_df_2022_1 = new_df_2022_1.drop(columns=['sum'])
    print('Công ty đầu tư năm ' +str(years_end)+' là ' ,new_df_2022_1[new_df_2022_1['diff in percentage'] == new_df_2022_1['diff in percentage'].max()]['SYMBOL'].values[0] )
    print('Tỷ lệ chêch lệch là ' ,new_df_2022_1[new_df_2022_1['diff in percentage'] == new_df_2022_1['diff in percentage'].max()]['diff in percentage'].values[0] , '%')
    return new_df_2022_1
final_table =find_company_to_invest(data,2007,2022)

Công ty đầu tư năm 2022 là  VSH
Tỷ lệ chêch lệch là  65.23054023801416 %


# Check lại trong dataframe nhận thấy giá trị PROFIT của mã VSH năm 2022 là 1.280645 chính xác với giá trị trong bảng formula_14_07_2023_16_15_00.csv

In [9]:
# FIND THE ROWS HAVE THE SYMBOL VSH IN 2022
filtered_data[filtered_data['SYMBOL'] == 'VSH'].head(1)

SYMBOL EXCHANGE  TIME    PROFIT     VALUEARG  Cash and cash equivalents  \
57    VSH     HOSE  2022  1.280645  978000000.0                   0.007155   

    Short Term Financial Investments  Short term receivables  \
57                          0.005363                0.067926   

    Total Inventories  Total Other Current Assets  ...  \
57           0.013916                    0.000001  ...   

    General and administration expenses  Net operating profit  Other income  \
57                             0.006193              0.062234      0.000012   

    Other expenses  Other profit  Total accounting profit before tax  \
57        0.000659     -0.000647                            0.061587   

    Current income tax  Deferred income tax  Profit after tax  ValueARG  
57            0.008706                  0.0          0.052881  0.000134  

[1 rows x 40 columns]

In [10]:
df.head(1)

formula  f_profit  TIME    profit
0       16  1.216969  2022  1.280645